nixos/modules/virtualisation/digital-ocean-init.nix

```diff
@@ -0,0 +1,63 @@
+{ config, pkgs, lib, ... }:
+with lib;
+{
+  options.virtualisation.digitalOcean.rebuildFromUserData = mkOption {
+    type = bool;
+    enable = true;
+    example = true;
+    description = "Whether to reconfigure the system from Digital Ocean user data";
+  };
+
+  config = {
+    systemd.services.digital-ocean-init = {
+      description = "Reconfigure the system from Digital Ocean uesrdata on startup";
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        ConditionFileExists = "!/etc/nixos/configuration.nix";
+        After = [ "network-online.target" ];
+        Wants = [ "network-online.target" ];
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+      restartIfChanged = false;
+      path = [ pkgs.curl pkgs.gnused pkgs.gnugrep pkgs.systemd config.nix.package config.system.build.nixos-rebuild ];
+      script = ''
+        set -e
+        echo "attempting to fetch configuration from Digital Ocean user data..."
+        export HOME=/root
+        export NIX_PATH=/nix/var/nix/profiles/per-user/root/channels/nixos:nixos-config=/etc/nixos/configuration.nix:/nix/var/nix/profiles/per-user/root/channels
+        userData=$(mktemp)
```

Comment:
I copied this from the Amazon init module and changed obvious things, but I haven't tested it yet.